File: C Coding/Searching And Sorting/heapsort.c
#include <stdio.h>
#include <stdlib.h>

void swap(int *a, int *b)
{
   int temp = *a;
   *a = *b;
   *b = temp;
}

void maxheapify(int arr[], int n, int i)
{
   int largest = i;
   int left = 2 * i + 1;
   int right = 2 * i + 2;

   if (left < n && arr[left] > arr[largest])
   {
      largest = left;
   }
   if (right < n && arr[right] > arr[largest])
   {
      largest = right;
   }

   if (largest != i)
   {
      swap(&arr[i], &arr[largest]);
      maxheapify(arr, n, largest);
   }
}
void buildmaxheap(int arr[], int n)
{
   for (int i = n / 2 - 1; i >= 0; i--)
   {
      maxheapify(arr,n,i);
   }
}

void heapsort(int arr[], int n){
   buildmaxheap(arr, n);
   for (int i = n-1; i >=1; i--)
   {
      swap(&arr[0], &arr[i]);
      maxheapify(arr, i , 0);
   }
   
}

void printArray(int arr[], int n)
{
   for (int i = 0; i < n; i++)
   {
      printf("%d ", arr[i]);
   }
   printf("\n");
}

int main()
{
   int arr[] = {9, 4, 3, 8, 10, 2, 5};
   int n = sizeof(arr) / sizeof(arr[0]);

   heapsort(arr, n);

   printf("Sorted array is : ");
   printArray(arr, n);
   return 0;

   return 0;
}

File: DSA/DAA/heapsort.java
package DAA;

import java.util.Arrays;

public class heapsort {
    // Heapify a subtree rooted with node i which is an index in arr[], n is size of heap
    private static void heapify(int[] arr, int n, int i) {
        int largest = i; // Initialize largest as root
        int left = 2 * i + 1; // left child
        int right = 2 * i + 2; // right child

        // If left child is larger than root
        if (left < n && arr[left] > arr[largest])
            largest = left;

        // If right child is larger than largest so far
        if (right < n && arr[right] > arr[largest])
            largest = right;

        // If largest is not root
        if (largest != i) {
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;

            // Recursively heapify the affected sub-tree
            heapify(arr, n, largest);
        }
    }

    // Main function to do heap sort
    public static void heapSort(int[] arr) {
        int n = arr.length;

        // Build heap (rearrange array)
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i);
        }
        System.out.println("After building max heap: " + Arrays.toString(arr));

        // One by one extract an element from heap
        for (int i = n - 1; i > 0; i--) {
            // Move current root to end
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;

            // call max heapify on the reduced heap
            heapify(arr, i, 0);
            System.out.println("After moving max to end and heapifying: " + Arrays.toString(arr));
        }
    }

    public static void main(String[] args) {
        int[] arr = {4, 10, 3, 5, 1};
        System.out.println("Original array: " + Arrays.toString(arr));
        heapSort(arr);
        System.out.println("Sorted array:   " + Arrays.toString(arr));
    }
}
